design_notes/v2.md

```diff
@@ -0,0 +1,173 @@
+`ManifestStore` - Contains a ManifestGroup and a FileReader
+`xarray.Dataset` - Contains ManifestArray-backed variables
+
+## Protocols
+
+> **Note**: Protocols over ABCs would allow third party backends to be less tightly coupled to VirtualiZarr
+> **Question:**: How do we simplify these protocol definitions using obspec?
+
+`FileReader` - read bytes from files on object storage or disk
+`VirtualZarrBackend` - construct a ManifestStore from a file
+
+```python!=
+class FileReader(protocol):
+    """
+    Protocol for file operations for downstream libraries
+    """
+
+    def read(self, size: int, /) -> bytes
+        ...
+
+    def seek(self, offset: int, whence: int = 0, /):
+        ...
+
+    def tell(self) -> int:
+        ...
+
+    async get_range_async(
+        path: str,
+        *,
+        start: int,
+        end: int | None = None,
+        length: int | None = None
+    ) -> Bytes :
+        ...
+
+    async get(
+        path: str,
+        *,
+        options: int | None = None,
+        length: int | None = None
+    ) -> GetResult :
+        ...
+
+```
+
+
+```python!=
+class VirtualZarrBackend(protocol):
+    """
+    Protocol for creating a virtual zarr dataset from a file
+    """
+
+    @staticmethod
+    def open_virtual_store(
+        filepath: str,
+        *,
+        virtual_backend_kwargs: dict | None = None,
+        reader_options: ObspecReaderOptions | dict | None = None,
+    ) → VirtualStore:
+        ...
+```
+
+> **Question:**: How do we define this protocol to allow for strict typing for some readers (e.g., obstore-backed) and not others (e.g., fsspec-backed)?
+
+> **Note**: Could ObspectReaderOptions by typed as follows:
+
+```python!=
+ObspecReaderOptions(TypedDict):
+    s3_config: S3Config | None = None,
+    azure_config: AzureConfig | None = None,
+    gcs_config: GCSConfig | None = None,
+    client_options: NotRequired[ClientConfig]
+    retry_config: NotRequired[RetryConfig]
+    credential_provider: NotRequired[Callable]
+```
+## Top-level API functions
+
+`open_virtual_dataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from a file
+`open_virtual_mfdataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from many files
+
+> **NOTE**: Here the construction of a virtual dataset from a manifest store would happen in the top-level function rather than the backends, which would need to implement `open_virtual_store` rather than `open_virtual_dataset`
+
+```python!=
+ def open_virtual_dataset(
+    filepath: str,
+    *,
+    cache: bool = False,
+    drop_variables: Iterable[str] | None = None,
+    loadable_variables: Iterable[str] | None = None,
+    decode_times: bool | None = None,
+    cftime_variables: Iterable[str] | None = None,
+    indexes: Mapping[str, Index] | None = None,
+    virtual_backend_kwargs: dict | None = None,
+    reader_options: ObspecReaderOptions | dict | None = None,
```

Comment:
This whole idea of passing in the `reader_options` is something we should avoid IMO, because it connects the public API into one specific implementation.

Reply:
Yes, I believe this ties in with my comment about allowing the user to pass in their own reader, which would also mean that the `cache` parameter goes away, as the user would simply pass in a caching reader, it they like.

Reply:
I like the idea of a user specified "reader" rather than the `reader_options` parameter and the removal of the `cache` parameter.

That said, I am not sure how we'd define what a `reader` is acknowledging the oddity above of having `async` and `sync` in a single protocol (https://github.com/zarr-developers/VirtualiZarr/pull/568#discussion_r2060688872) and that different backends require different methods. For example, h5py-based backends require `read`, `seek` and `tell`. But others might not (e.g., `async_tiff`, maybe?). The ManifestStore used for all loadable variables requires only `get_range_async `. Would the "reader" be used for both, or would we have a "object_reader" and a "file_reader" as separate parameters, or some other solution?

Reply:
@maxrjones, I'm not familiar enough with all of this, but based upon the ["caching tweak" you recently created](https://github.com/zarr-developers/VirtualiZarr/blob/d6d6c25e76ac412555b06f91eec9d8548fc5f72d/virtualizarr/readers/hdf/hdf.py), I suspect what we might actually want to allow the user to pass in their own _store_ rather than their own _reader_, and that store could be a "caching" store.

I'm only basing this suggestion off or my recent gleaning of that "tweak" you created, so I could be missing something in how all these pieces fit together, so if my suggestion doesn't make sense, please add some information to help me better grasp everything.

Reply:
> what we might actually want to allow the user to pass in their own store rather than their own reader, and that store could be a "caching" store.

So the `CachingStore` would simply cache the entire file if any part of it is asked for? That sounds neat.